File: JavaHTTPRecipesEdition1/com/heatonresearch/httprecipes/ch13/recipe1/LinkReport.java
package com.heatonresearch.httprecipes.ch13.recipe1;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.logging.*;

import com.heatonresearch.httprecipes.spider.*;
import com.heatonresearch.httprecipes.spider.workload.*;

/**
 * Recipe #13.1: Scanning for broken links 
 * Copyright 2007 by Jeff Heaton(jeff@jeffheaton.com)
 *
 * HTTP Programming Recipes for Java Bots
 * ISBN: 0-9773206-6-9
 * http://www.heatonresearch.com/articles/series/16/
 * 
 * The report class for the Link Checker.
 * 
 * This software is copyrighted. You may use it in programs
 * of your own, without restriction, but you may not publish
 * the source code without the author's permission. For more
 * information on distributing this code, please visit:
 * http://www.heatonresearch.com/hr_legal.php
 * 
 * @author Jeff Heaton
 * @version 1.1
 */
public class LinkReport implements SpiderReportable {
  /**
   * For logging.
   */
  private static Logger logger = Logger
      .getLogger("com.heatonresearch.httprecipes.spider.Spider");

  /**
   * The host we are working with.
   */
  private String base;

  /**
   * The bad URL's.
   */
  private List<String> bad = new ArrayList<String>();

  /**
   * The workload manager being used.
   */
  private WorkloadManager workloadManager;

  /**
   * This function is called when the spider is ready to
   * process a new host. This function simply stores the
   * value of the current host.
   * 
   * @param host
   *          The new host that is about to be processed.
   * @return True if this host should be processed, false
   *         otherwise.
   */
  public boolean beginHost(String host) {
    if (this.base == null) {
      this.base = host;
      return true;
    } else {
      return false;
    }
  }

  /**
   * @return the bad
   */
  public List<String> getBad() {
    return this.bad;
  }

  /**
   * Called when the spider is starting up. This method
   * provides the SpiderReportable class with the spider
   * object.
   * 
   * @param spider
   *          The spider that will be working with this
   *          object.
   */
  public void init(Spider spider) {
    this.workloadManager = spider.getWorkloadManager();
  }

  /**
   * @param bad
   *          the bad to set
   */
  public void setBad(List<String> bad) {
    this.bad = bad;
  }

  public boolean spiderFoundURL(URL url, URL source,
      SpiderReportable.URLType type) {

    if ((this.base != null) && (!this.base.equalsIgnoreCase(url.getHost()))) {
      return false;
    }

    return true;
  }

  /**
   * Called when the spider is about to process a NON-HTML
   * URL.
   * 
   * @param url
   *          The URL that the spider found.
   * @param stream
   *          An InputStream to read the page contents from.
   * @throws IOException
   *           Thrown if an IO error occurs while processing
   *           the page.
   */
  public void spiderProcessURL(URL url, InputStream stream) {
  }

  /**
   * Called when the spider is ready to process an HTML
   * URL.
   * 
   * @param url
   *          The URL that the spider is about to process.
   * @param parse
   *          An object that will allow you you to parse the
   *          HTML on this page.
   * @throws IOException
   *           Thrown if an IO error occurs while processing
   *           the page.
   */
  public void spiderProcessURL(URL url, SpiderParseHTML parse) {
    try {
      parse.readAll();
    } catch (IOException e) {
      logger.log(Level.INFO, "Error reading page:" + url.toString());
    }
  }

  /**
   * Called when the spider tries to process a URL but gets
   * an error.
   * 
   * @param url
   *          The URL that generated an error.
   */
  public void spiderURLError(URL url) {
    URL source;
    try {
      source = this.workloadManager.getSource(url);
      StringBuilder str = new StringBuilder();
      str.append("Bad URL:");
      str.append(url.toString());
      str.append(" found at ");
      str.append(source.toString());
      this.bad.add(str.toString());
    } catch (WorkloadException e) {
      e.printStackTrace();
    }
  }

}

File: CSharpHTTPRecipes/HTTPRecipes/Recipe13_1/LinkReport.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using HeatonResearch.Spider;
using HeatonResearch.Spider.Workload;
using HeatonResearch.Spider.Logging;

namespace Recipe13_1
{
    /// <summary>
    /// Recipe #13.1: Scanning for broken links 
    /// Copyright 2007 by Jeff Heaton(jeff@jeffheaton.com)
    ///
    /// HTTP Programming Recipes for C# Bots
    /// ISBN: 0-9773206-7-7
    /// http://www.heatonresearch.com/articles/series/20/
    ///
    /// The report class for the Link Checker.
    ///
    /// This software is copyrighted. You may use it in programs
    /// of your own, without restriction, but you may not
    /// publish the source code without the author's permission.
    /// For more information on distributing this code, please
    /// visit:
    ///    http://www.heatonresearch.com/hr_legal.php
    /// </summary>
    class LinkReport : SpiderReportable
    {
        /// <summary>
        /// The host we are working with.
        /// </summary>
        private String baseHost;

        /// <summary>
        /// The Spider object that this object reports to.
        /// </summary>
        private Spider spider;

        /// <summary>
        /// The bad URL's.
        /// </summary>
        private List<String> bad = new List<String>();

        /// <summary>
        /// This function is called when the spider is ready to
        /// process a new host. This function simply stores the
        /// value of the current host.
        /// </summary>
        /// <param name="host"></param>
        /// <returns></returns>
        public bool BeginHost(String host)
        {
            if (this.baseHost == null)
            {
                this.baseHost = host;
                return true;
            }
            else
            {
                return false;
            }
        }

        /// <summary>
        /// The bad link's found.
        /// </summary>
        public List<String> Bad
        {
            get
            {
                return this.bad;
            }
            set
            {
                bad = value;
            }
        }

        /// <summary>
        /// Called when the spider is starting up. This method
        /// provides the SpiderReportable class with the spider
        /// object.
        /// </summary>
        /// <param name="spider">The spider that will be working with this object.</param>
        public void Init(Spider spider)
        {
            this.spider = spider;
        }


        /// <summary>
        /// Called when the spider finds a URL.
        /// </summary>
        /// <param name="url">The URL that was found.</param>
        /// <param name="source">Where the URL was found.</param>
        /// <param name="type">What sort of tag produced this URL.</param>
        /// <returns></returns>
        public bool SpiderFoundURL(Uri url, Uri source,
            Spider.URLType type)
        {
            if ((this.baseHost != null) && (String.Compare(this.baseHost, url.Host, true) != 0))
            {
                return false;
            }

            return true;
        }

        /// <summary>
        /// Not used by the link checker.
        /// </summary>
        /// <param name="url">Not used.</param>
        /// <param name="stream">Not used.</param>
        public void SpiderProcessURL(Uri url, Stream stream)
        {
        }

        /// <summary>
        /// Called when the spider is ready to process an HTML
        /// URL.
        /// </summary>
        /// <param name="url">The URL that the spider is about to process.</param>
        /// <param name="parse">An object that will allow you you to parse the HTML on this page.</param>
        public void SpiderProcessURL(Uri url, SpiderParseHTML parse)
        {
            try
            {
                parse.ReadAll();
            }
            catch (IOException)
            {
                spider.Logging.Log(Logger.Level.INFO, "Error reading page:" + url.ToString());
            }
        }

        /// <summary>
        /// Called when the spider tries to process a URL but gets
        /// an error.
        /// </summary>
        /// <param name="url">The URL that generated an error.</param>
        public void SpiderURLError(Uri url)
        {
            Uri source;
            try
            {
                source = this.spider.Workload.GetSource(url);
                StringBuilder str = new StringBuilder();
                str.Append("Bad URL:");
                str.Append(url.ToString());
                str.Append(" found at ");
                str.Append(source.ToString());
                this.bad.Add(str.ToString());
            }
            catch (WorkloadException e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }
    }
}
